<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Targeted Learning with Moderated Statistics for Biomarker Discovery • biotmle</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Targeted Learning with Moderated Statistics for Biomarker Discovery">
<meta property="og:description" content="This package facilitates the discovery of biomarkers from
    biological sequencing data (e.g., microarrays, RNA-seq) based on the
    associations of potential biomarkers with exposure and outcome variables by
    implementing an estimation procedure that combines a generalization of
    moderated statistics with targeted minimum loss-based estimates (TMLE) of
    parameters defined via causal inference (e.g., Average Treatment Effect)
    whose estimators admit asymptotically linear representations.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">biotmle</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/exposureBiomarkers.html">Identifying Biomarkers from an Exposure Variable</a>
    </li>
    <li>
      <a href="articles/rnaseqProcessing.html">Processing and Analyzing RNA-Seq Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nhejazi/biotmle">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="rbiotmle" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rbiotmle" class="anchor"></a>R/<code>biotmle</code>
</h1></div>

<blockquote>
<p>Targeted Learning with Moderated Statistics for Biomarker Discovery</p>
</blockquote>
<p><strong>Author:</strong> <a href="https://nimahejazi.org">Nima Hejazi</a></p>
<hr>
<div id="whats-biotmle" class="section level2">
<h2 class="hasAnchor">
<a href="#whats-biotmle" class="anchor"></a>What’s <code>biotmle</code>?</h2>
<p><code>biotmle</code> is an R package that facilitates biomarker discovery by generalizing the moderated t-statistic <span class="citation">(Smyth 2004)</span> for use with target parameters that have asymptotically linear representations <span class="citation">(van der Laan and Rose 2011)</span>. The set of methods implemented rely on the use of targeted maximum likelihood (TML) estimation to transform biological sequencing data (e.g., microarray, RNA-seq) based on the (efficient) influence function (EIF) representation of a particular causal target parameter (e.g., average treatment effect). The transformed data (rotated into influence function space) are then be subjected to a moderated test for differences between the statistical estimate of the target parameter and a hypothesized value of said parameter (usually a null value defined in relation to the parameter itself). Such an approach provides a valid and conservative statistical hypothesis test of a statistically estimable target parameter while controlling the standard error (derived from the variance of the EIF) such that the error rate of the test is more strongly controlled relative to testing procedures that do fail moderate the variance estimate <span class="citation">(<span class="citeproc-not-found" data-reference-id="hejazi2018variance"><strong>???</strong></span>)</span> while simultaneously avoiding many of the pitfalls that plague analogous procedures predicated upon correctly specifying linear models <span class="citation">(Smyth 2004)</span>. Utilities are also provided for performing clustering based on <em>supervised distance matrices</em>, using the EIF-based estimates to draw out underlying contributions of biomarkers to the target parameter of interest <span class="citation">(Pollard and van der Laan 2008)</span>.</p>
<hr>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>For standard use, install from <a href="https://bioconductor.org/packages/biotmle">Bioconductor</a> using <a href="https://CRAN.R-project.org/package=BiocManager"><code>BiocManager</code></a>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly=</span><span class="ot">TRUE</span>)) {</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">  <span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">}</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">BiocManager<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"biotmle"</span>)</a></code></pre></div>
<p>To contribute, install the bleeding-edge <em>development version</em> from GitHub via <a href="https://www.rstudio.com/products/rpackages/devtools/"><code>devtools</code></a>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"nhejazi/biotmle"</span>)</a></code></pre></div>
<p>Current and prior <a href="https://bioconductor.org">Bioconductor</a> releases are available under branches with numbers prefixed by “RELEASE_”. For example, to install the version of this package available via Bioconductor 3.6, use</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"nhejazi/biotmle"</span>, <span class="dt">ref =</span> <span class="st">"RELEASE_3_6"</span>)</a></code></pre></div>
<hr>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>For details on how to best use the <code>biotmle</code> R package, please consult the most recent <a href="https://bioconductor.org/packages/release/bioc/vignettes/biotmle/inst/doc/exposureBiomarkers.html">package vignette</a> available through the <a href="https://bioconductor.org/packages/biotmle">Bioconductor project</a>.</p>
<hr>
</div>
<div id="issues" class="section level2">
<h2 class="hasAnchor">
<a href="#issues" class="anchor"></a>Issues</h2>
<p>If you encounter any bugs or have any specific feature requests, please <a href="https://github.com/nhejazi/biotmle/issues">file an issue</a>.</p>
<hr>
</div>
<div id="contributions" class="section level2">
<h2 class="hasAnchor">
<a href="#contributions" class="anchor"></a>Contributions</h2>
<p>Contributions are very welcome. Interested contributors should consult our <a href="https://github.com/nhejazi/biotmle/blob/master/CONTRIBUTING.html">contribution guidelines</a> prior to submitting a pull request.</p>
<hr>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>After using the <code>biotmle</code> R package, please cite both of the following:</p>
<pre><code>    @article{hejazi2017biotmle,
      author = {Hejazi, Nima S and Cai, Weixin and Hubbard, Alan E},
      title = {biotmle: Targeted Learning for Biomarker Discovery},
      journal = {The Journal of Open Source Software},
      volume = {2},
      number = {15},
      month = {July},
      year  = {2017},
      publisher = {The Open Journal},
      doi = {10.21105/joss.00295},
      url = {https://doi.org/10.21105/joss.00295}
    }

    @article{hejazi2018+variance,
      url = {https://arxiv.org/abs/1710.05451},
      year = {2018+},
      author = {Hejazi, Nima S and {Kherad-Pajouh}, Sara and {van der
        Laan}, Mark J and Hubbard, Alan E},
      title = {Variance moderation of locally efficient estimators in
        high-dimensional biology}
    }</code></pre>
<hr>
</div>
<div id="related" class="section level2">
<h2 class="hasAnchor">
<a href="#related" class="anchor"></a>Related</h2>
<ul>
<li>
<a href="https://github.com/nhejazi/biotmleData">R/<code>biotmleData</code></a> - R package with example experimental data for use with this analysis package.</li>
</ul>
<hr>
</div>
<div id="funding" class="section level2">
<h2 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h2>
<p>The development of this software was supported in part through grants from the National Institutes of Health: <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9260357&amp;map=y">P42 ES004705-29</a> and <a href="https://projectreporter.nih.gov/project_info_description.cfm?aid=9210551&amp;icde=37849782&amp;ddparam=&amp;ddvalue=&amp;ddsub=&amp;cr=1&amp;csb=default&amp;cs=ASC&amp;pball=">R01 ES021369-05</a>.</p>
<hr>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>License</h2>
<p>© 2016-2018 <a href="https://nimahejazi.org">Nima S. Hejazi</a></p>
<p>The contents of this repository are distributed under the MIT license. See file <code>LICENSE</code> for details.</p>
<hr>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-pollard2008supervised">
<p>Pollard, Katherine S, and Mark J van der Laan. 2008. “Supervised Distance Matrices.” <em>Statistical Applications in Genetics and Molecular Biology</em> 7 (1). De Gruyter.</p>
</div>
<div id="ref-smyth2004linear">
<p>Smyth, Gordon K. 2004. “Linear Models and Empirical Bayes Methods for Assessing Differential Expression in Microarray Experiments.” <em>Statistical Applications in Genetics and Molecular Biology</em> 3 (1). Walter de Gruyter: 1–25. <a href="https://doi.org/10.2202/1544-6115.1027" class="uri">https://doi.org/10.2202/1544-6115.1027</a>.</p>
</div>
<div id="ref-vdl2011targeted">
<p>van der Laan, Mark J., and Sherri Rose. 2011. <em>Targeted Learning: Causal Inference for Observational and Experimental Data</em>. Springer Science &amp; Business Media.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/biotmle">https://​www.bioconductor.org/​packages/​biotmle</a>
</li>
<li>Browse source code at <br><a href="https://github.com/nhejazi/biotmle">https://​github.com/​nhejazi/​biotmle</a>
</li>
<li>Report a bug at <br><a href="https://github.com/nhejazi/biotmle/issues">https://​github.com/​nhejazi/​biotmle/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing biotmle</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Nima Hejazi <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-7127-2789" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li>Alan Hubbard <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0002-3769-0127" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/nhejazi/biotmle"><img src="https://travis-ci.org/nhejazi/biotmle.svg?branch=master" alt="Travis-CI Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/nhejazi/biotmle/"><img src="https://ci.appveyor.com/api/projects/status/github/nhejazi/biotmle?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://codecov.io/github/nhejazi/biotmle?branch=master"><img src="https://img.shields.io/codecov/c/github/nhejazi/biotmle/master.svg" alt="Coverage Status"></a></li>
<li><a href="http://www.repostatus.org/#active"><img src="http://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/biotmle"><img src="http://www.bioconductor.org/shields/build/release/bioc/biotmle.svg" alt="BioC status"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/biotmle.html"><img src="http://bioconductor.org/shields/years-in-bioc/biotmle.svg" alt="Bioc Time"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/biotmle.html"><img src="http://bioconductor.org/shields/downloads/biotmle.svg" alt="Bioc Downloads"></a></li>
<li><a href="http://opensource.org/licenses/MIT"><img src="http://img.shields.io/badge/license-MIT-brightgreen.svg" alt="MIT license"></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/65854775"><img src="https://zenodo.org/badge/65854775.svg" alt="DOI"></a></li>
<li><a href="http://joss.theoj.org/papers/02be843d9bab1b598187bfbb08ce3949"><img src="http://joss.theoj.org/papers/02be843d9bab1b598187bfbb08ce3949/status.svg" alt="JOSS Status"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nima Hejazi, Alan Hubbard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
