<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Biomarker Evaluation with Targeted Minimum Loss-Based Estimation of the ATE — biomarkertmle • biotmle</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Biomarker Evaluation with Targeted Minimum Loss-Based Estimation of the ATE — biomarkertmle" />

<meta property="og:description" content="Computes the causal target parameter defined as the difference between the
biomarker expression values under treatment and those same values under no
treatment, using Targeted Minimum Loss-Based Estimation." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biotmle</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/exposureBiomarkers.html">Identifying Biomarkers from an Exposure Variable</a>
    </li>
    <li>
      <a href="../articles/rnaseqProcessing.html">Processing and Analyzing RNA-Seq Data</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nhejazi/biotmle">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Biomarker Evaluation with Targeted Minimum Loss-Based Estimation of the ATE</h1>
    <small class="dont-index">Source: <a href='https://github.com/nhejazi/biotmle/blob/master/R/biomarkertmle-wrapper.R'><code>R/biomarkertmle-wrapper.R</code></a></small>
    <div class="hidden name"><code>biomarkertmle.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Computes the causal target parameter defined as the difference between the
biomarker expression values under treatment and those same values under no
treatment, using Targeted Minimum Loss-Based Estimation.</p>
    
    </div>

    <pre class="usage"><span class='fu'>biomarkertmle</span>(<span class='no'>se</span>, <span class='no'>varInt</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ngscounts</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>bppar_type</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>future_param</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>family</span> <span class='kw'>=</span> <span class='st'>"gaussian"</span>, <span class='kw'>subj_ids</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>g_lib</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"SL.glm"</span>,
  <span class='st'>"SL.randomForest"</span>, <span class='st'>"SL.nnet"</span>, <span class='st'>"SL.polymars"</span>, <span class='st'>"SL.mean"</span>),
  <span class='kw'>Q_lib</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"SL.glm"</span>, <span class='st'>"SL.randomForest"</span>, <span class='st'>"SL.nnet"</span>, <span class='st'>"SL.mean"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>se</th>
      <td><p>(SummarizedExperiment) - containing expression or next-generation
sequencing data in the "assays" slot and a matrix of phenotype-level data
in the "colData" slot.</p></td>
    </tr>
    <tr>
      <th>varInt</th>
      <td><p>(numeric) - indicating the column of the design matrix
corresponding to the treatment or outcome of interest (in the
<code>colData</code> slot of the <code>SummarizedExperiment</code> argument "se").</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p>(logical) - whether the data included in the <code>assay</code>
slot of the input <code>SummarizedExperiment</code> object has been normalized
already. The default is set to <code>TRUE</code> since it is expected that most
practitioners would apply normalization methods appropriate to the type of
assay being analyzed. If set to <code>FALSE</code>, median normalization is
performed for microarray (i.e., non-RNA-seq) data.</p></td>
    </tr>
    <tr>
      <th>ngscounts</th>
      <td><p>(logical) - whether the data are counts generated from a
next-generation sequencing (NGS) experiment (e.g., RNA-seq). The default
setting assumes continuous expression measures as generated by platforms
that are microarray-type (i.e., so-called "targeted" assays).</p></td>
    </tr>
    <tr>
      <th>parallel</th>
      <td><p>(logical) - whether or not to use parallelization in the
estimation procedure. Invoking parallelization happens through a
combination of calls to <code>future</code> and <code>BiocParallel</code>. If this
argument is set to <code>TRUE</code>, <code><a href='http://www.rdocumentation.org/packages/future/topics/multiprocess'>future::multiprocess</a></code> is used,
and if <code>FALSE</code>, <code><a href='http://www.rdocumentation.org/packages/future/topics/sequential'>future::sequential</a></code> is used, alongside
<code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/bplapply'>BiocParallel::bplapply</a></code>. Other options for evaluation through
futures may be invoked by setting the argument <code>future_param</code>.</p></td>
    </tr>
    <tr>
      <th>bppar_type</th>
      <td><p>(character) - specifies the type of backend to be used with
the parallelization invoked by <code>BiocParallel</code>. Consult the manual page
for <code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/BiocParallelParam-class'>BiocParallel::BiocParallelParam</a></code> for possible types and
descriptions on their appropriate uses. The default for this argument is
<code>NULL</code>, which silently uses <code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/DoparParam-class'>BiocParallel::DoparParam</a></code>.</p></td>
    </tr>
    <tr>
      <th>future_param</th>
      <td><p>(character) - specifies the type of parallelization to be
invoked when using futures for evaluation. For a list of the available
types, please consult the documentation for <code><a href='http://www.rdocumentation.org/packages/future/topics/plan'>future::plan</a></code>. The default
setting (this argument set to <code>NULL</code>) silently invokes
<code><a href='http://www.rdocumentation.org/packages/future/topics/multiprocess'>future::multiprocess</a></code>. Be careful if changing this setting.</p></td>
    </tr>
    <tr>
      <th>family</th>
      <td><p>(character) - specification of error family: "binomial" or
"gaussian".</p></td>
    </tr>
    <tr>
      <th>subj_ids</th>
      <td><p>(numeric vector) - subject IDs to be passed directly to
subject should have the exact same numerical identifier; coerced to class
<code>numeric</code> if not provided in the appropriate form.</p></td>
    </tr>
    <tr>
      <th>g_lib</th>
      <td><p>(char vector) - library of learning algorithms to be used in
fitting the propensity score (the nuisance parameter denoted "g" in the
literature on targeted minimum loss-based estimation).</p></td>
    </tr>
    <tr>
      <th>Q_lib</th>
      <td><p>(char vector) - library of learning algorithms to be used in
fitting the outcome regression (the nuisance parameter denoted "Q" in the
literature on targeted minimum loss-based estimation).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>S4 object of class <code>biotmle</code>, generated by sub-classing
 <code>SummarizedExperiment</code>, with additional slots containing <code>tmleOut</code>
 and <code>call</code>, among others, containing TMLE-based estimates of the
 relationship between a biomarker and exposure or outcome variable and the
 original call to this function (for user reference), respectively.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>dplyr</span>)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘dplyr’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    count</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:Biobase’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    combine</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:GenomicRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    intersect, setdiff, union</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:GenomeInfoDb’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    intersect</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:IRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    collapse, desc, intersect, setdiff, slice, union</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:S4Vectors’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    first, intersect, rename, setdiff, setequal, union</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:BiocGenerics’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    combine, intersect, setdiff, union</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:biotmle’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    combine, intersect, setdiff, union</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:testthat’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    matches</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    filter, lag</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    intersect, setdiff, setequal, union</span></div><div class='input'><span class='fu'>library</span>(<span class='no'>biotmleData</span>)
<span class='fu'>data</span>(<span class='no'>illuminaData</span>)
<span class='fu'>library</span>(<span class='no'>SummarizedExperiment</span>)
<span class='st'>"%ni%"</span> <span class='kw'>=</span> <span class='fu'>Negate</span>(<span class='st'>"%in%"</span>)

<span class='fu'><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>colData</a></span>(<span class='no'>illuminaData</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>colData</a></span>(<span class='no'>illuminaData</span>) <span class='kw'>%&gt;%</span>
     <span class='no'>data.frame</span> <span class='kw'>%&gt;%</span>
     <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='http://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span>(<span class='kw'>age</span> <span class='kw'>=</span> <span class='fu'>as.numeric</span>(<span class='no'>age</span> <span class='kw'>&gt;</span> <span class='fu'>median</span>(<span class='no'>age</span>))) <span class='kw'>%&gt;%</span>
     <span class='no'>DataFrame</span>

<span class='no'>varInt_index</span> <span class='kw'>&lt;-</span> <span class='fu'>which</span>(<span class='fu'>names</span>(<span class='fu'><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>colData</a></span>(<span class='no'>illuminaData</span>)) <span class='kw'>%in%</span> <span class='st'>"benzene"</span>)

<span class='no'>biomarkerTMLEout</span> <span class='kw'>&lt;-</span> <span class='fu'>biomarkertmle</span>(<span class='kw'>se</span> <span class='kw'>=</span> <span class='no'>illuminaData</span>[<span class='fl'>1</span>:<span class='fl'>2</span>, ],
                                  <span class='kw'>varInt</span> <span class='kw'>=</span> <span class='no'>varInt_index</span>,
                                  <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
                                  <span class='kw'>family</span> <span class='kw'>=</span> <span class='st'>"gaussian"</span>,
                                  <span class='kw'>g_lib</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"SL.mean"</span>, <span class='st'>"SL.glm"</span>),
                                  <span class='kw'>Q_lib</span> <span class='kw'>=</span> <span class='st'>"SL.mean"</span>
                                 )</div><div class='output co'>#&gt; <span class='warning'>Warning: Sequential evaluation is strongly discouraged. </span>
#&gt; <span class='warning'> Proceed with caution.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: nnls</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Nima Hejazi, Alan Hubbard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

